src/control/lib/hardware/topology.go

```diff
@@ -34,20 +41,130 @@ func (t *Topology) AllDevices() map[string]*Device {
 	return devsByName
 }
 
+func PrintTopology(t *Topology, output io.Writer) error {
+	if t == nil {
+		fmt.Fprintf(output, "No topology information available\n")
+	}
+
+	for _, numaNode := range t.NUMANodes {
+		coreSet := &system.RankSet{}
+		for _, core := range numaNode.Cores {
+			coreSet.Add(system.Rank(core.ID))
+		}
+
+		fmt.Fprintf(output, "NUMA node %d (%s)\n", numaNode.ID, numaNode.Set)
+		fmt.Fprintf(output, "  CPU cores: %s\n", coreSet)
+		fmt.Fprintf(output, "  PCI devices: \n")
+		for addr, devs := range numaNode.PCIDevices {
+			fmt.Fprintf(output, "    %s\n", &addr)
+			for _, dev := range devs {
+				fmt.Fprintf(output, "      %s\n", dev)
+			}
+		}
+		fmt.Fprintf(output, "  PCI buses: \n")
+		for _, bus := range numaNode.PCIBuses {
+			fmt.Fprintf(output, "    %s\n", bus)
+		}
+	}
+
+	return nil
+}
+
 type (
+	CPUCore struct {
+		ID       uint      `json:"id"`
+		NUMANode *NUMANode `json:"-"`
+	}
```

Comment:
A bit curious about the use case for knowing the core ID. Something to do with assigning the engine threads to cores?

Reply:
Not exactly sure yet, but yes, thinking about enabling that sort of thing. We also eventually should be trying to pin the control plane processes to specific cores in order to avoid contention with I/O. If nothing else, I just figured that the information is there, so it can't hurt to gather it.